spec-examples/src/blake2/blake2b.rs

```diff
@@ -49,51 +73,75 @@ fn mix(v: DoubleState, a: usize, b: usize, c: usize, d: usize, x: U64, y: U64) -
     result
 }
 
-// TODO: add test case where counter wraps
-fn inc_counter(t: Counter, x: u64) -> Counter {
-    let mut result: Counter = Counter([0; 2]);
-    result[0] = t[0] + x;
-    if result[0] < x {
-        result[1] = t[1] + 1u64;
+fn inc_counter<Word: PublicInteger>(t: Counter<Word>, x: Word) -> Counter<Word> {
+    let mut result = Counter::new();
+    let new_val: Word = t[0] + x;
+    result[0] = new_val;
+    if new_val.less_than(x) {
+        result[1] = t[1] + Word::ONE;
     }
     result
 }
 
-// TODO: move to library
-fn make_u64array(h: Buffer) -> DoubleState {
+fn make_array<Word: UnsignedSecretInteger>(h: &ByteSeq) -> DoubleState<Word> {
+    assert_eq!(h.len() / ((Word::NUM_BITS as usize) / 8), 16);
     let mut result = DoubleState::new();
     for i in 0..16 {
-        result[i] = U64::from(h[8 * i])
-            | U64::from(h[1 + 8 * i]) << 8
-            | U64::from(h[2 + 8 * i]) << 16
-            | U64::from(h[3 + 8 * i]) << 24
-            | U64::from(h[4 + 8 * i]) << 32
-            | U64::from(h[5 + 8 * i]) << 40
-            | U64::from(h[6 + 8 * i]) << 48
-            | U64::from(h[7 + 8 * i]) << 56;
+        let (_, h_chunk) = h.get_chunk(Word::NUM_BITS as usize / 8, i);
+        result[i] = Word::from_le_bytes(&h_chunk)
     }
     result
 }
 
-fn compress(h: State, m: Buffer, t: Counter, last_block: bool) -> State {
+pub trait HasIV<Word: UnsignedSecretInteger> {
```

Comment:
I'm really not happy having a trait in here. But let's keep it and see how we can improve on this.

Reply:
Yeah I didn't find better